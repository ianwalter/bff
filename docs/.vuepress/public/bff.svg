<?xml version="1.0" encoding="UTF-8"?>
<svg width="200px" height="200px" viewBox="0 0 200 200" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 54.1 (76490) - https://sketchapp.com -->
    <title>Letter</title>
    <desc>Created with Sketch.</desc>
    <defs>
        <linearGradient x1="0%" y1="7.9854923%" x2="100%" y2="92.0145077%" id="linearGradient-1">
            <stop stop-color="#30E8BF" offset="0%"></stop>
            <stop stop-color="#38EF7D" offset="100%"></stop>
        </linearGradient>
    </defs>
    <g id="Letter" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="iconmonstr-favorite-11" transform="translate(28.000000, 34.000000)" fill="url(#linearGradient-1)" fill-rule="nonzero">
            <path d="M117,54 C102.08925,54 90,66.08925 90,81 C90,95.91075 102.08925,108 117,108 C131.91075,108 144,95.91075 144,81 C144,66.08925 131.91075,54 117,54 Z M114.1875,92.907 L104.0625,83.088 L108.24075,78.9075 L114.1875,84.546 L126.88425,71.532 L131.0625,75.71025 L114.1875,92.907 Z M92.838,111.558 C86.19,118.074 79.188,124.872 72,132 C33.42,93.714 0,65.118 0,37.152 C0,-3.462 53.112,-12.924 72,19.488 C90.75,-12.684 144,-3.6 144,37.152 C144,41.628 143.064,46.128 141.462,50.67 C134.766,45.258 126.258,42 117,42 C95.496,42 78,59.496 78,81 C78,93.378 83.82,104.406 92.838,111.558 Z" id="Shape"></path>
        </g>
    </g>
</svg>