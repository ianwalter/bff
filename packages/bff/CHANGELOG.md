# @ianwalter/bff

## 11.2.2

### Patch Changes

- 9ed47c9: Update dependency @generates/plug to ^1.0.2
- a15e4b6: Update dependency @generates/logger to ^2.0.4

## 11.2.1

### Patch Changes

- ad375bb: Update dependency @generates/logger to ^2.0.3
- d616507: Update dependency @generates/plug to ^1.0.1
- e567f1c: Update dependency common-tags to ^1.8.2

## 11.2.0

### Minor Changes

- e9ae633: Update dependency jest-snapshot to ^27.3.1

### Patch Changes

- f97e4d6: Update dependency @generates/plug to v1

## 11.1.1

### Patch Changes

- f8b3a8a: Update dependency @generates/plug to ^0.1.1
- 05f0686: Update dependency @generates/logger to v2
- 77a62fd: Update dependency expect to ^27.2.5
- 8ba795a: Update dependency @generates/cli to v2
- 6c375e6: Update dependency jest-snapshot to ^27.2.5
- effa4f1: Update dependency glob to ^7.2.0

## 11.1.0

### Minor Changes

- a2f9064: Update dependency expect to v27

### Patch Changes

- d382186: Update dependency workerpool to ^6.1.5
- 37a4601: Update dependency array-shuffle to v3
- c688c77: Update dependency jest-snapshot to v27

## 11.0.1

### Patch Changes

- c674f38: Fix #420: Dont show skipped tests when running with the --failed flag
- e17caeb: Update dependency callsites to v4
- acbc85f: Update dependency glob to ^7.1.7

## 11.0.0

### Major Changes

- f455e82: Replacing plugin system with @generates/plug

## 10.2.0

### Minor Changes

- 1e10635: Add #390: Add logger in enhanceTestContext

## 10.1.0

### Minor Changes

- 8cb3368: Add #367: Add t.warn and t.skip to enhanceTestContext
- 0df5bc8: Add #383: Enhance test context before "each" hooks are called
- 3c51612: Updating cli to use @generates/cli
- 158dd03: Running after hook on SIGINT

### Patch Changes

- 51e7ebf: Update dependency @generates/logger to v1
- 7676ac2: Update dependency webdriverio to ^7.3.1

## 10.0.7

### Patch Changes

- b6187ef: Update dependency workerpool to ^6.1.3

## 10.0.6

### Patch Changes

- 23472d8: Update dependency junit-report-builder to v3
- 8f35c44: Update dependency @generates/logger to ^0.1.2
- 1e592c4: Update dependency @generates/logger to ^0.1.1

## 10.0.5

### Patch Changes

- 6cd3a0f: Update dependency camaro to ^6.1.0
- 59d6bd3: Update dependency workerpool to ^6.1.0
- 80f5fce: Fixing generates/logger import

## 10.0.4

### Patch Changes

- 0b8cc04: Fix #341: dont show skipped test line numbers with --verbose

## 10.0.3

### Patch Changes

- 38660a5: Fixing #324: file globbing

## 10.0.2

### Patch Changes

- 9420121: Update dependency workerpool to ^6.0.4
- 4d4b31b: Update dependency @generates/logger to ^0.1.0
- 4bd38f4: Update dependency @generates/merger to ^0.1.3

## 10.0.1

### Patch Changes

- a3f105d: Updating @generates/merger

## 10.0.0

### Major Changes

- 3a243c9: Converting to ESM

## 9.13.0

### Minor Changes

- 605e72f: Adding support for ESM

## 9.12.3

### Patch Changes

- 330baa7: Cleaning up logger changes and bff-webdriver integration API

## 9.12.2

### Patch Changes

- 7c59efd: Update dependency workerpool to ^6.0.3
- cb49d76: Update dependency expect to ^26.6.0
- 9d61389: Update dependency execa to v5
- 5b562e9: Update dependency array-shuffle to v2
- 5aaf6e3: Update dependency jest-snapshot to ^26.6.2
- 7c3a041: Update dependency jest-snapshot to ^26.6.0
- 71a533a: Update dependency expect to ^26.6.2

## 9.12.1

### Patch Changes

- e643d40: Fixing / improving SIGINT logic

## 9.12.0

### Minor Changes

- 5d86cd2: Updating logger

## 9.11.0

### Minor Changes

- 389ec6e: Update dependency expect to ^26.5.3
- 4420311: Update dependency jest-snapshot to ^26.5.3

### Patch Changes

- 6605982: Update dependency workerpool to ^6.0.2
