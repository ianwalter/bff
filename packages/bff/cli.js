#!/usr/bin/env node

import { promises as fs } from 'fs'
import path from 'path'
import cli from '@generates/cli'
import generatesLogger from '@generates/logger'
import camaro from 'camaro'
import junitBuilder from 'junit-report-builder'
import * as bff from './index.js'

// Set stdout to blocking so that the program doesn't exit with log statements
// still waiting to be logged to the console.
if (process.stdout._handle) process.stdout._handle.setBlocking(true)

const { createLogger, chalk } = generatesLogger
const logger = createLogger({ namespace: 'bff.cli', level: 'info' })

async function run () {
  const input = cli({
    name: 'bff',
    usage: 'bff [path-to-tests] [options]',
    options: {
      concurrency: {
        aliases: ['c'],
        arg: '<number>',
        description: 'Specifies how many tests/workers to run in parallel'
      },
      updateSnapshot: {
        aliases: ['u'],
        description: 'Specifies whether snapshots should be created or updated',
        default: false
      },
      log: {
        aliases: ['l'],
        description: 'Specifies logging configuration',
        default: { namespace: 'bff.main', level: 'info' }
      },
      tag: {
        aliases: ['t'],
        arg: '<tag>',
        description: `
          Specifies which test tags should be used to match tests. How it
          matches the tags depends on the \`match\` option below
        `
      },
      timeout: {
        aliases: ['T'],
        arg: '<milliseconds>',
        description: `
          Specifies how long a test should take in milliseconds before it's
          marked as failed for timing out
        `,
        default: 60000
      },
      failFast: {
        aliases: ['f'],
        description: `
          Specifies whether to exit when a test fails instead of continuing to
          run tests
        `,
        default: false
      },
      junit: {
        aliases: ['j'],
        arg: '[path]',
        description: `
          Specifies whether or not to write the results to a junit report file
          and optionally the relative path of the file
        `
      },
      match: {
        aliases: ['m'],
        arg: '<type>',
        description: `
          Specifies whether a test needs \`some\` or \`every\` specified tag in
          order for it to be run
        `,
        default: 'some'
      },
      verbose: {
        aliases: ['V'],
        description: `
          Logs more information for each test: test tags, relative file path,
          and timing information
        `,
        default: false
      },
      runs: {
        aliases: ['r'],
        arg: '<count>',
        description: 'Specifies the number of test runs to execute',
        default: 1
      },
      failed: {
        aliases: ['F'],
        arg: '[junit file]',
        description: `
          Only run tests marked as failed in ./junit.xml (or specified file)
        `
      }
    }
  })

  if (input?.helpText) {
    process.stdout.write('\n')
    logger.info(input.helpText)
    process.stdout.write('\n')
    process.exit(0)
  }

  // Only run tests marked as failed in a JUnit file.
  if (input.failed) {
    const file = typeof input.failed === 'string' ? input.failed : 'junit.xml'
    const xml = await fs.readFile(path.resolve(file), 'utf8')
    const template = { failed: ['//testcase[failure]', '@name'] }
    const { failed } = await camaro.transform(xml, template)
    process.stdout.write('\n')
    logger.info(`Running failed tests in ${file}:`, '\n', failed.join('\n'))
    process.stdout.write('\n')
    input.failed = failed
  }

  // Set tests as whatever paths were passed as input to the CLI or whatever
  // is configured and delete the args attribute to get rid of duplicate data.
  input.tests = input.args.length ? input.args : input.tests
  delete input.args

  // Run the tests and wait for a response with the passed/failed/skipped
  // counts.
  const passed = []
  const failed = []
  const warnings = []
  const skipped = []
  for (let runs = 0; runs < input.runs; runs++) {
    const result = await bff.run(input)

    // Add a blank line between the test output and result summary so it's
    // easier to spot.
    process.stdout.write('\n')

    // If there was an error thrown outside of the test functions (e.g.
    // requiring a module that wasn't found) then output a fatal error.
    if (result.err) {
      logger.fatal(result.err)
      if (result.err instanceof bff.FailFastError) {
        process.stdout.write('\n')
      } else {
        process.exit(1)
      }
    }

    // Log the results of running the tests.
    logger.info(
      chalk.green.bold(`${result.passed.length} passed.`),
      chalk.red.bold(`${result.failed.length} failed.`),
      chalk.yellow.bold(`${result.warnings.length} warnings.`),
      chalk.white.bold(`${result.skipped.length} skipped.`)
    )

    // Aggregate test results accross runs.
    passed.push(...result.passed)
    failed.push(...result.failed)
    warnings.push(...result.warnings)
    skipped.push(...result.skipped)

    // Add blank line after the result summary so it's easier to spot.
    process.stdout.write('\n')
  }

  // If configured, generate a junit XML report file based on the test results.
  if (input.junit) {
    // Determine the junit report file path.
    const junit = typeof input.junit === 'string' ? input.junit : 'junit.xml'

    // Group tests by test file so that the test file relative path can be used
    // as the suite name.
    const allTests = [...passed, ...failed, ...warnings, ...skipped]
    const files = allTests.reduce((acc, test) => {
      if (acc[test.file]) {
        acc[test.file].push(test)
      } else {
        acc[test.file] = [test]
      }
      return acc
    }, {})

    // Create a test for each test file and add the containing tests to the
    // suite as test cases.
    Object.entries(files).forEach(([file, tests]) => {
      const suite = junitBuilder.testSuite().name(file)
      tests.forEach(test => {
        const testCase = suite.testCase().name(test.name)
        if (test.skip || (test.err && test.warn)) {
          testCase.skipped()
        } else if (test.err) {
          testCase.failure(test.err)
        }
      })
    })

    // Write the junit report file to the filesystem.
    junitBuilder.writeTo(path.resolve(junit))
  }

  // If any tests failed, exit with a non-zero exit code.
  process.exit(failed.length ? 1 : 0)
}

run().catch(err => {
  process.stdout.write('\n')
  logger.fatal(err)
  process.exit(1)
})
