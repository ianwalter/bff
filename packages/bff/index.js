import readline from 'readline'
import path from 'path'
import util from 'util'
import { fileURLToPath } from 'url'
import workerpool from 'workerpool'
import glob from 'glob'
import generatesLogger from '@generates/logger'
import { oneLine } from 'common-tags'
import pSeries from 'p-series'
import jestSnapshot from 'jest-snapshot'
import { merge } from '@generates/merger'
import callsites from 'callsites'
import shuffle from 'array-shuffle'
import toHookRun from './lib/toHookRun.js'

const { createLogger, chalk } = generatesLogger
const { SnapshotState } = jestSnapshot
const globa = util.promisify(glob)
const defaultFiles = [
  '*@(tests|play|pptr).?(m|c)js',
  'tests/**/*@(tests|play|pptr).?(m|c)js'
]

export class FailFastError extends Error {
  constructor () {
    super(FailFastError.message)
  }
}
FailFastError.message = 'Run failed immediately since failFast option is set'

/**
 * Collects test names from test files and assigns them to a worker in a
 * worker pool that runs the associated test.
 */
export async function run (config) {
  // Create the run context using the passed configuration and defaults.
  const context = {
    tests: defaultFiles,
    testContext: { hasRun: false, result: {}, timeout: config.timeout },
    plugins: [],
    // Initialize a count for each time a test file has been registered so that
    // the main thread can figure out when registration has completed and the
    // worker pool can be terminated.
    filesRegistered: 0,
    // Initialize a count for the total number of tests registered from all of
    // the test files.
    testsRegistered: 0,
    // Initialize collections for tests that passed, failed, or were skipped.
    passed: [],
    failed: [],
    warnings: [],
    skipped: [],
    // Initialize a count for the total number of tests that have been run so
    // that the run can figure out when all tests have completed and the worker
    // pool can be terminated.
    testsRun: 0,
    enhanceTestContext: true
  }

  // Destructure passed configuration and add it to testContext and context.
  const { updateSnapshot, tag = [], failed, ...restOfConfig } = config
  context.testContext.updateSnapshot = updateSnapshot ? 'all' : 'none'
  context.tags = Array.isArray(tag) ? tag : [tag]
  merge(context, restOfConfig)

  // Create the logger instance with the given log level.
  const logger = createLogger(context.log)

  // Add the absolute paths of the test files to the run context.
  const ignore = 'node_modules/**'
  const globOptions = { nosort: true, nodir: true, ignore, absolute: true }
  const files = await Promise.all(context.tests.map(t => globa(t, globOptions)))
  context.files = shuffle(files.flat())
  logger.debug('Run context', context)

  // Throw an error if there are no tests files found.
  if (context.files.length === 0) throw new Error('No test files found.')

  // Set the worker pool options. For now, it only sets the maximum amount of
  // workers used if the concurrency setting is set.
  const poolOptions = {
    ...(context.concurrency ? { maxWorkers: context.concurrency } : {})
  }

  // Set the path to the file used to create a worker.
  const __dirname = path.dirname(fileURLToPath(import.meta.url))
  const workerPath = path.join(__dirname, 'worker.js')

  // For registering individual tests exported from test files:
  const registrationPool = workerpool.pool(workerPath, poolOptions)

  // For actually running the tests:
  const runPool = workerpool.pool(workerPath, poolOptions)

  // Create readline instance so bff can listen for multiple SIGINT events.
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  })

  // Handle <ctrl>c / SIGINT events.
  rl.on('SIGINT', async function onSigint () {
    // Inform the user that the event has been received.
    process.stdout.write('\n')
    if (context.receivedSigint) {
      // Terminate the workers immediately.
      logger.warn('Second SIGINT received. Forcing worker termination.')
      return runPool.terminate(true)
    } else {
      logger.warn('SIGINT received. Forwarding to workers.')
    }
    process.stdout.write('\n')

    // Terminate the registration workers immediately.
    registrationPool.terminate(true)

    // If SIGINT wasn't already received, run after hooks.
    if (!context.receivedSigint) {
      await pSeries(context.plugins.map(toHookRun('after', context)))
    }

    // Keep track of the fact that bff has received a SIGINT.
    context.receivedSigint = true

    // Mark the run as having failed in the context.
    context.err = new Error('RUN CANCELLED!')

    // Forward the SIGINT to the test workers via the seppuku task.
    for (const worker of runPool.workers) worker.exec('seppuku')
  })

  // Sequentially run any before hooks specified by plugins.
  await pSeries(context.plugins.map(toHookRun('before', context)))

  try {
    // For each test file found, pass the filename to a registration pool
    // worker so that the tests within it can be collected and given to a
    // run pool worker to be run.
    await Promise.all(context.files.map(async filePath => {
      // Create the file context to contain information on the test file.
      const relativePath = path.relative(process.cwd(), filePath)
      const file = { path: filePath, relativePath }

      // Construct the path to the snapshot file.
      const snapshotsDir = path.join(path.dirname(filePath), 'snapshots')
      const snapshotFilename = `${path.basename(filePath)}.snap`
      file.snapshotPath = path.join(snapshotsDir, snapshotFilename)

      // Perform registration on the test file to collect the tests that need
      // to be run.
      merge(file, await registrationPool.exec('register', [file, context]))

      // Increment the registration count now that registration has completed
      // for the current test file.
      context.filesRegistered++

      // Terminate the registration pool if all the test files have been
      // registered.
      if (context.filesRegistered === context.files.length) {
        const numberOfTests = context.testsRegistered + file.tests.length
        logger.debug('Number of tests:', numberOfTests)
        registrationPool.terminate()
          .then(() => logger.debug('Registration pool terminated'))
      }

      // Add the number of tests returned by test registration to the running
      // total of all tests that need to be run.
      context.testsRegistered += file.tests.length

      // Determine if any of the tests in the test file have the .only
      // modifier so that tests can be excluded from being run.
      const hasOnly = Object.values(file.tests).some(test => test.only)

      // Get the snapshot state for the current test file.
      const snapshotState = new SnapshotState(
        file.snapshotPath,
        { updateSnapshot: context.testContext.updateSnapshot }
      )

      // Iterate through all tests in the test file.
      await Promise.all(shuffle(file.tests).map(async test => {
        let result
        try {
          // Mark all tests as having been checked for snapshot changes so
          // that tests that have been removed can have their associated
          // snapshots removed as well when the snapshots are checked for this
          // test file.
          snapshotState.markSnapshotsAsCheckedForTest(test.name)

          const skipViaOnly = hasOnly && !test.only
          if (skipViaOnly || test.skip) {
            // Output the test name and increment the skip count to remind
            // the user that some tests are being explicitly skipped.
            const msg = `${context.testsRun + 1}. ${test.name}`
            logger.log('ðŸ›Œ', msg, skipViaOnly ? chalk.dim('(via only)') : '')
            context.skipped.push({ ...test, file: relativePath })
          } else if (!failed || failed.includes(test.name)) {
            // Send the test to a worker in the run pool to be run.
            result = await runPool.exec('test', [file, test, context])
            logger.debug('Test result', { test: test.name, ...result })

            // If t.skip was called within the test, mark it as skipped.
            if (result.skipped) {
              logger.log('ðŸ›Œ', `${context.testsRun + 1}. ${test.name}`)
              return context.skipped.push({ ...test, file: relativePath })
            }

            // If t.warn was called within the test, mark it as a warning.
            if (result.warned) {
              logger.warn(`${context.testsRun + 1}. ${test.name}`)
              return context.warnings.push({ ...test, file: relativePath })
            }

            // Update the snapshot state with the snapshot data received from
            // the worker.
            if (result && (result.added || result.updated)) {
              snapshotState._dirty = true
              snapshotState._counters = new Map(result.counters)
              Object.assign(snapshotState._snapshotData, result.snapshots)
              snapshotState.added += result.added
              snapshotState.updated += result.updated
            }

            // Output the test name and increment the pass count since the
            // test didn't throw an error indicating a failure.
            logger.success(`${context.testsRun + 1}. ${test.name}`)
            context.passed.push({ ...test, file: relativePath })
          }
        } catch (err) {
          const file = relativePath
          const workerpoolErrors = ['Worker terminated', 'Pool terminated']
          if (workerpoolErrors.includes(err.message)) {
            // Ignore 'Worker terminated' errors since there is already output
            // when a run is cancelled.
            return
          } if (test.warn) {
            logger.warn(`${context.testsRun + 1}. ${test.name}:`, err)
            return context.warnings.push({ ...test, err: err.message, file })
          } else if (err.name === 'TimeoutError') {
            const msg = `${context.testsRun + 1}. ${test.name}: timeout`
            logger.error(msg, chalk.dim(file))
          } else {
            logger.error(`${context.testsRun + 1}. ${test.name}:`, err)
          }

          // Increment the failure count since the test threw an error
          // indicating a test failure.
          context.failed.push({ ...test, err: err.message, file })
        } finally {
          // Increment the test run count now that the test has completed.
          context.testsRun++

          // Log the relative file path and test duration if in verbose mode.
          if (context.verbose && !context.receivedSigint && result) {
            const pad = ''.padEnd((context.testsRun * 100).toString().length)
            logger.log(`${pad}${file.relativePath}:${test.lineNumber}`)
            if (result.duration) {
              logger.log(chalk.dim(`${pad}in`, result.duration))
            }
          }
        }

        // If the failFast option is set, throw an error so that the test run is
        // immediately failed.
        const [err] = context.failed
        if (err && context.failFast) throw new FailFastError()
      }))

      // The snapshot tests that weren't checked are obsolete and can be
      // removed from the snapshot file.
      if (snapshotState.getUncheckedCount()) snapshotState.removeUncheckedKeys()

      // Save the snapshot changes.
      snapshotState.save()
    }))
  } catch (err) {
    // Add the fatal error to the context so the CLI can fail the run.
    context.err = err
  }

  // Sequentially run any after hooks specified by plugins.
  await pSeries(context.plugins.map(toHookRun('after', context)))

  // Terminate the run pool now that all tests have been run.
  runPool.terminate().then(() => logger.debug('Run pool terminated'))

  return context
}

function handleTestArgs (name, tags, test = {}) {
  // Add the test line number to the object so it can be shown in verbose mode.
  test.lineNumber = callsites()[2].getLineNumber()

  // Extract the test function from the function arguments.
  const testFn = tags.pop()
  Object.assign(test, { fn: testFn, tags })
  const key = oneLine(name)

  // Add the test to the global namespaced by the test filename.
  const file = global.bff.file
  if (!global.bff.tests[file]) global.bff.tests[file] = {}
  global.bff.tests[file][key] = test

  if (testFn && typeof testFn === 'function') {
    return test
  } else {
    return fn => {
      Object.assign(test, { fn, tags: testFn ? [...tags, testFn] : [] })
      return test
    }
  }
}

export function test (name, ...tags) {
  return handleTestArgs(name, tags)
}

test.skip = function skip (name, ...tags) {
  return handleTestArgs(name, tags, { skip: true })
}

test.only = function only (name, ...tags) {
  return handleTestArgs(name, tags, { only: true })
}

test.warn = function warn (name, ...tags) {
  return handleTestArgs(name, tags, { warn: true })
}
