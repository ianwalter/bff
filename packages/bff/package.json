{
  "name": "@ianwalter/bff",
  "description": "Your friendly test framework",
  "version": "11.1.1",
  "license": "SEE LICENSE IN LICENSE",
  "type": "module",
  "module": "index.js",
  "scripts": {
    "test": "./cli.js",
    "test.ci": "./cli.js -V tests/bff.js",
    "test.out": "./cli.js tests/stdout.js | wc -c"
  },
  "bin": {
    "bff": "cli.js"
  },
  "files": [
    "lib",
    "cli.js",
    "index.js",
    "worker.js"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ianwalter/bff.git",
    "directory": "packages/bff"
  },
  "keywords": [
    "testing",
    "test",
    "runner",
    "framework",
    "tdd"
  ],
  "author": {
    "name": "Ian Walter",
    "url": "https://ianwalter.dev"
  },
  "bugs": {
    "url": "https://github.com/ianwalter/bff/issues"
  },
  "funding": {
    "type": "github",
    "url": "https://github.com/sponsors/ianwalter"
  },
  "homepage": "https://github.com/ianwalter/bff#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@generates/cli": "^2.0.0",
    "@generates/logger": "^2.0.2",
    "@generates/merger": "^0.1.3",
    "@generates/plug": "^1.0.0",
    "@ianwalter/clone": "^6.0.1",
    "@ianwalter/cloneable": "^2.0.0",
    "@ianwalter/sleep": "^0.0.1",
    "@ianwalter/subpub": "^3.0.0",
    "@ianwalter/timer": "^2.0.2",
    "array-shuffle": "^3.0.0",
    "callsites": "^4.0.0",
    "camaro": "^6.1.0",
    "common-tags": "^1.8.0",
    "esm-to-cjs": "^1.2.0",
    "expect": "^27.2.5",
    "glob": "^7.2.0",
    "jest-snapshot": "^27.3.1",
    "junit-report-builder": "^3.0.0",
    "workerpool": "^6.1.5"
  },
  "devDependencies": {
    "execa": "^5.1.1"
  }
}
