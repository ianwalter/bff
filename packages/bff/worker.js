const path = require('path')
const { worker } = require('workerpool')
const pSeries = require('p-series')
const { createLogger, chalk } = require('@generates/logger')

let threadId = process.pid
try {
  const workerThreads = require('worker_threads')
  threadId = workerThreads.threadId
} catch (err) {
  // Ignore error.
}

worker({
  seppuku () {
    // Emit a SIGINT to itself so that processes terminate gracefully.
    process.kill(process.pid, 'SIGINT')
  },
  async register (file, context) {
    // Create the logger instance based on the log level set in the context
    // received from the main thread.
    const namespace = `bff.worker.${threadId}.register`
    const logger = createLogger({ ...context.log, namespace })

    // Log a debug statement for this registration action with the relative
    // path of the test file that's having it's tests registered.
    const relativePath = chalk.dim(file.relativePath)
    logger.debug(`Registration worker ${threadId}`, relativePath)

    // Sequentially run any registration hooks specified by plugins.
    const toHookRun = require('./lib/toHookRun')
    if (context.plugins && context.plugins.length) {
      await pSeries(
        context.plugins.map(toHookRun('registration', file, context))
      )
    }

    // If the map of tests in the current test file hasn't been added to the
    // context, require the test file and use it's exports object as the test
    // map.
    if (!context.testMap) {
      try {
        context.testMap = require(file.path)
      } catch (err) {
        if (err.code === 'ERR_REQUIRE_ESM') {
          const dist = require('@ianwalter/dist')
          const requireFromString = require('require-from-string')
          const { cjs } = await dist({ input: file.path, cjs: true })
          requireFromString(cjs[1], file.name)
          context.testMap = global.tests
        } else {
          throw err
        }
      }
    }

    // Add a list of tests from the test file that are intended to be run to
    // the file context.
    const { tags, match } = context
    const tagsMatch = test => {
      if (['some', 'every'].includes(match)) {
        return tags[match](tag => test.tags.includes(tag))
      }
      throw new Error(`match value must be 'some' or 'every', not '${match}'`)
    }
    file.tests = Object.entries(context.testMap).reduce(
      (acc, [name, test]) => !tags.length || (tags.length && tagsMatch(test))
        ? acc.concat([{ key: name, name, ...test, fn: null }])
        : acc,
      []
    )

    // If an augmentTests method has been added to the context by a plugin, call
    // it with the list of tests so that the plugin can alter them if necessary.
    if (context.augmentTests) {
      file.tests = context.augmentTests(file.tests)
    }

    // Return the file context with the the list of registered tests.
    return file
  },
  async test (file, test, context) {
    const merge = require('@ianwalter/merge')
    const createTimer = require('@ianwalter/timer')
    const toHookRun = require('./lib/toHookRun')

    // Create the logger instance based on the log level set in the context
    // received from the main thread.
    const namespace = `bff.worker.${threadId}.test`
    const logger = createLogger({ ...context.log, namespace })

    // Log a debug statement for this test action with the test name and
    // relative path of the test file it belongs to.
    const relativePath = chalk.dim(file.relativePath)
    logger.debug(`Test worker ${threadId}`, chalk.cyan(test.name), relativePath)

    // Add the file and test data to the testContext.
    merge(context.testContext, file, test)

    try {
      if (context.plugins && context.plugins.length) {
        // Sequentially run any beforeEach hooks specified by plugins.
        await pSeries(
          context.plugins.map(toHookRun('beforeEach', file, context))
        )

        // If the verbose option is set, start a timer for the test.
        if (context.verbose) {
          context.timer = createTimer()
        }

        // Sequentially run any runTest hooks specified by plugins.
        await pSeries(context.plugins.map(toHookRun('runTest', file, context)))
      }

      if (!context.testContext.hasRun) {
        // If the verbose option is set, start a timer for the test.
        if (context.verbose) {
          context.timer = createTimer()
        }

        // Enhance the context passed to the test function with testing
        // utilities.
        const enhanceTestContext = require('./lib/enhanceTestContext')
        enhanceTestContext(context.testContext)
        context.testContext.logger = logger

        try {
          // TODO: Load the test file and
          require(file.path)
        } catch (err) {
          if (err.code === 'ERR_REQUIRE_ESM') {
            const dist = require('@ianwalter/dist')
            const requireFromString = require('require-from-string')
            const { cjs } = await dist({ input: file.path, cjs: true })
            requireFromString(cjs[1], file.name)
          } else {
            throw err
          }
        }

        // Run the test!
        const runTest = require('./lib/runTest')
        await runTest(context.testContext, global.tests[test.key].fn)
        context.testContext.hasRun = true
      }

      // If there was a timer started for the test, stop the timer, get the
      // timer's duration, and add it to the test result.
      if (context.timer) {
        const duration = context.timer.duration()
        logger.debug('Test duration', duration)
        context.testContext.result.duration = duration
      }
    } finally {
      // Sequentially run any afterEach hooks specified by plugins.
      if (context.plugins && context.plugins.length) {
        await pSeries(
          context.plugins.map(toHookRun('afterEach', file, context))
        )
      }
    }

    // Return the test result to the main thread.
    if (context.testContext.result.failed) {
      throw context.testContext.result.failed
    }
    return context.testContext.result
  }
})
