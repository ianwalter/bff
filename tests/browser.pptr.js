import { test } from '@ianwalter/bff'

test('manual pass', ({ pass }) => pass())

test('arithmetic', ({ expect }) => expect(13 + 8).toBe(21))

test('dom snapshot', ({ expect }) => {
  const anchor = document.createElement('a')
  anchor.setAttribute('href', 'https://iankwalter.com')
  anchor.innerHTML = `Ian's Website`
  expect(anchor).toMatchSnapshot()
})
