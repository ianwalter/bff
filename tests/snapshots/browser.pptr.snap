// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dom snapshot 1`] = `
<a
  href="https://iankwalter.com"
>
  Ian's Website
</a>
`;
